File: src/main/java/com/github/mgramin/sqlboot/model/resource_type/wrappers/list/CacheWrapper.java
/**
 * The MIT License (MIT)
 *
 * Copyright (c) 2016-2017 Maksim Gramin
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.github.mgramin.sqlboot.model.resource_type.wrappers.list;

import static java.util.Optional.ofNullable;

import com.github.mgramin.sqlboot.exceptions.BootException;
import com.github.mgramin.sqlboot.model.resource.DbResource;
import com.github.mgramin.sqlboot.model.resource_type.ResourceType;
import com.github.mgramin.sqlboot.model.uri.Uri;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import javax.cache.Cache;
import javax.cache.CacheManager;
import javax.cache.Caching;
import javax.cache.configuration.MutableConfiguration;
import javax.cache.spi.CachingProvider;

/**
 * @author Maksim Gramin (mgramin@gmail.com)
 * @version $Id$
 * @since 0.1
 */
public class CacheWrapper implements ResourceType {

    private final ResourceType origin;

    private final CachingProvider cachingProvider;
    private final CacheManager cacheManager;
    private final MutableConfiguration<String, List<DbResource>> config;
    private /*final*/ Cache<String, List<DbResource>> cache;

    public CacheWrapper(ResourceType origin) {
        cachingProvider = Caching.getCachingProvider();
        cacheManager = cachingProvider.getCacheManager();
        config = new MutableConfiguration<>();
        cache = cacheManager.getCache("simpleCache");
        if (cache == null) {
            cache = cacheManager.createCache("simpleCache", config);
        }
        this.origin = origin;
    }

    @Override
    public List<String> aliases() {
        return origin.aliases();
    }

    @Override
    public List<String> path() {
        return origin.path();
    }

    @Override
    public Stream<DbResource> read(Uri uri) throws BootException {
        final String cache = ofNullable(uri.params().get("cache")).orElse("true");
        List<DbResource> cachedResources = this.cache.get(uri.toString());
        if (cachedResources == null || cache.equalsIgnoreCase("false")) {
            cachedResources = origin.read(uri).collect(Collectors.toList());
            this.cache.put(uri.toString(), cachedResources);
        }
        return cachedResources.stream();
    }

    @Override
    public Map<String, String> metaData() {
        return origin.metaData();
    }

}

File: src/main/kotlin/com/github/mgramin/sqlboot/model/resourcetype/wrappers/list/CacheWrapper.kt
/*
 * BSD 3-Clause License
 *
 * Copyright (c) 2019, CROC Inc.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the copyright holder nor the names of its
 *    contributors may be used to endorse or promote products derived from
 *    this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

package com.github.mgramin.sqlboot.model.resourcetype.wrappers.list

import com.github.mgramin.sqlboot.exceptions.BootException
import com.github.mgramin.sqlboot.model.resource.DbResource
import com.github.mgramin.sqlboot.model.resourcetype.Metadata
import com.github.mgramin.sqlboot.model.resourcetype.ResourceType
import com.github.mgramin.sqlboot.model.uri.Uri
import reactor.core.publisher.Flux
import reactor.kotlin.core.publisher.toFlux
import javax.cache.Cache
import javax.cache.CacheManager
import javax.cache.Caching
import javax.cache.configuration.MutableConfiguration

/**
 * @author Maksim Gramin (mgramin@gmail.com)
 * @version $Id: 822c72ab4745f06ca5b3062b4b0be1f9588596db $
 * @since 0.1
 */
class CacheWrapper(private val origin: ResourceType, private val parameterName: String = "cache") : ResourceType {

    private val cacheManager: CacheManager = Caching.getCachingProvider().cacheManager
    private val cache: Cache<String, List<DbResource>> =
            cacheManager.getCache("simpleCache")
                    ?: cacheManager.createCache("simpleCache", MutableConfiguration())

    override fun aliases(): List<String> {
        return origin.aliases()
    }

    override fun path(): List<String> {
        return origin.path()
    }

    @Throws(BootException::class)
    override fun read(uri: Uri): Flux<DbResource> {
        val cache = uri.params()[parameterName] ?: "true"
        var cachedResources: List<DbResource>? = this.cache.get(uri.toString())
        if (cachedResources == null || cache.equals("false", ignoreCase = true)) {
            cachedResources = origin.read(uri).collectList().block()
            this.cache.put(uri.toString(), cachedResources)
        }
        return cachedResources!!.toFlux()
    }

    override fun metaData(uri: Uri): List<Metadata> {
        return origin.metaData(uri)
    }

    override fun toJson() = origin.toJson()
}
